src/System.Management.Automation/engine/parser/PSType.cs

```diff
@@ -436,15 +437,29 @@ private Type GetBaseTypes(Parser parser, TypeDefinitionAst typeDefinitionAst, ou
 
             private bool ShouldImplementProperty(string name, Type type)
             {
-                foreach (var interfaceType in _typeBuilder.GetInterfaces())
+                if (_interfaceProperties == null)
                 {
-                    if (interfaceType.GetProperty(name, type) != null)
+                    _interfaceProperties = new HashSet<Tuple<string, Type>>();
+                    foreach (var interfaceType in _typeBuilder.GetInterfaces())
                     {
-                        return true;
+                        // TypeBuilder.GetInterfaces() returns only the interfaces that was explicitly passed to its constructor.
+                        // During compilation the interface hierarchy is flattened, so we only need to resolve one level of ancestral interfaces.
+                        foreach (var parentInterface in interfaceType.GetInterfaces())
+                        {
+                            foreach (var property in parentInterface.GetProperties())
+                            {
+                                _interfaceProperties.Add(new Tuple<string, Type>(property.Name, property.PropertyType));
+                            }
+                        }
+
+                        foreach (var property in interfaceType.GetProperties())
+                        {
+                            _interfaceProperties.Add(new Tuple<string, Type>(property.Name, property.PropertyType));
+                        }
                     }
                 }
 
-                return false;
+                return _interfaceProperties.Contains(new Tuple<string, Type>(name, type));
```

Comment:
nit: maybe use `Tuple.Create(name, type)` to replace `new Tuple<string, Type>(name, type)`.